File: Example/App_cbuilder/Unit1.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "Unit1.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TForm1 *Form1;
//---------------------------------------------------------------------------
__fastcall TForm1::TForm1(TComponent* Owner)
        : TForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TForm1::Button1Click(TObject *Sender)
{
   OpenDialog1->Filter = "DD(*.dll)|*.dll";
    OpenDialog1->DefaultExt = ".dll";
         if (this->OpenDialog1->Execute()) {
	this->Edit1->Text = OpenDialog1->FileName;
        AnsiString strM; strM = Edit1->Text;
		hinst = LoadLibraryW(this->Edit1->Text.w_str());

        this->Button2->Enabled = false;


		if (!hinst) {
		ShowMessage("Load Dll error");
	    return;
	}

	dd_todc = (lpfun_DD_todc)GetProcAddress(hinst, "DD_todc");
	dd_btn = (lpfun_DD_btn)GetProcAddress(hinst, "DD_btn");
	dd_mov = (lpfun_DD_mov)GetProcAddress(hinst, "DD_mov");
	dd_movR = (lpfun_DD_mov)GetProcAddress(hinst, "DD_movR");
	dd_whl = (lpfun_DD_whl)GetProcAddress(hinst, "DD_whl");
	dd_key = (lpfun_DD_key)GetProcAddress(hinst, "DD_key");
	dd_str = (lpfun_DD_str)GetProcAddress(hinst, "DD_str");

        if (dd_todc && dd_movR && dd_btn && dd_mov && dd_whl && dd_key && dd_str)
	{
		int st = dd_btn(0); //DD Initialize
		if(st == 1){
			this->Button2->Enabled = true;
			ShowMessage("OK");
        }
	}
	else {
		ShowMessage("Error");
	}

        }

}
//---------------------------------------------------------------------------

void __fastcall TForm1::Button2Click(TObject *Sender)
{
if (this->Button2->Caption == "Start") {
	this->Button2->Caption = "Stop";
	this->Timer1->Enabled = true;
    }
    else {
	this->Button2->Caption = "Start";
	this->Timer1->Enabled = false;
    }
}
//---------------------------------------------------------------------------


void __fastcall TForm1::FormClose(TObject *Sender, TCloseAction &Action)
{
                    if (this->Timer1->Enabled) {
	this->Timer1->Enabled = false;
    }
    if (hinst) {
	FreeLibrary(hinst);
    }
}
//---------------------------------------------------------------------------


void __fastcall TForm1::Timer1Timer(TObject *Sender)
{

    int ret =0;

  switch (select) {

	case 1:
	   {
		int ddcode = 300;            // TAB is 300 in ddcode.
		ddcode = dd_todc(VK_TAB);    //or by VK code

		dd_key(ddcode, 1);           //1==down£¬2==up
		Sleep(50);					 //may, delay 1-120ms,
		dd_key(ddcode, 2);
	   }
		case 2:
		dd_btn(1);         //1==L.down, 2==L.up, 4==R.down, 8==R.up, 16==M.down, 32==M.up
		Sleep(8);
		dd_btn(2);
	break;
	//dd_movR(10, 10);     //Mouse Move rel.
	//dd_mov(100, 100);    //Mouse Move abs.
	//ret = dd_whl(1);     //Mouse Wheel. 1==down, 2==up
	//ret = dd_whl(2);
	   }
}
//---------------------------------------------------------------------------

void __fastcall TForm1::RadioButton1Click(TObject *Sender)
{
  select=1;
}
//---------------------------------------------------------------------------

void __fastcall TForm1::RadioButton2Click(TObject *Sender)
{
  select=2;
}
//---------------------------------------------------------------------------

void __fastcall TForm1::FormCreate(TObject *Sender)
{
  select=0;
}
//---------------------------------------------------------------------------

File: Example/App_delphi/Unit1.pas
unit Unit1;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  Dialogs, StdCtrls, ExtCtrls, ExtDlgs;

type
  TForm1 = class(TForm)
    Timer1: TTimer;
    GroupBox1: TGroupBox;
    RadioButton1: TRadioButton;
    RadioButton2: TRadioButton;
    RadioButton3: TRadioButton;
    RadioButton4: TRadioButton;
    Button2: TButton;
    Button1: TButton;
    Edit1: TEdit;
    Button3: TButton;
    OpenDialog1: TOpenDialog;
    procedure FormCreate(Sender: TObject);
    procedure FormClose(Sender: TObject; var Action: TCloseAction);
    procedure Button2Click(Sender: TObject);
    procedure RadioButton1Click(Sender: TObject);
    procedure RadioButton2Click(Sender: TObject);
    procedure RadioButton3Click(Sender: TObject);
    procedure RadioButton4Click(Sender: TObject);

    procedure Timer1Timer(Sender: TObject);
    procedure Button1Click(Sender: TObject);

    procedure Button3Click(Sender: TObject);
    procedure GetDDFunAddress();

  private
    { Private declarations }

  public
    { Public declarations }
  end;

  //DD Function
  TDD_todc = function(vkcode:integer):integer;stdcall;
  TDD_mov = function(x,y:integer):integer;stdcall;
  TDD_btn = function(btn:integer):integer;stdcall;
  TDD_whl = function(whl:integer):integer;stdcall;
  TDD_key = function(key,flag:integer):integer;stdcall;
  TDD_str = function(str:AnsiString):integer;stdcall;
  TDD_movR = function(dx,dy:integer):integer;stdcall;
var
  Form1: TForm1;

var
  DD:Thandle;

  DD_todc:TDD_todc;
  DD_mov:TDD_mov;
  DD_btn:TDD_btn;
  DD_whl:TDD_whl;
  DD_key:TDD_key;
  DD_str:TDD_str;
  DD_movR:TDD_movR;

  select:integer;
implementation

{$R *.dfm}

procedure TForm1.FormCreate(Sender: TObject);

begin
  Button2.Enabled:=false;
  Button1.Enabled:=false;
end;

procedure TForm1.FormClose(Sender: TObject; var Action: TCloseAction);
begin
  if DD  <> 0 then
  begin
    freelibrary(DD );
  end;

  Timer1.Enabled := false;

end;

procedure TForm1.Button2Click(Sender: TObject);
begin
   Timer1.Enabled := false;
   Button1.Enabled:=true;
   Button2.Enabled:=false;
end;

procedure TForm1.Button3Click(Sender: TObject);
var
  filename:PChar;

begin
  if OpenDialog1.Execute then
  begin
    filename:= PChar(OpenDialog1.FileName);
    Edit1.Text:=filename;

    DD:=Loadlibrary( filename);  //x86 -> *32.dll , x64 -> *.64.dll

    if (DD  <> 0)  then
    begin
        MessageBox(0,'Load OK','info',MB_ICONASTERISK);
        GetDDFunAddress();
    end
    else
    begin
        MessageBox(0,'Load Error','info',MB_ICONASTERISK);
    end;
  end;
end;

procedure TForm1.RadioButton1Click(Sender: TObject);
begin
 select:=1;
end;

procedure TForm1.RadioButton2Click(Sender: TObject);
begin
 select:=2;
end;

procedure TForm1.RadioButton3Click(Sender: TObject);
begin
 select:=3;
end;

procedure TForm1.RadioButton4Click(Sender: TObject);
begin
 select:=4;
end;

procedure TForm1.Timer1Timer(Sender: TObject);
var
  ddcode :integer;
begin
   case select of
     1:
     begin
       DD_movR(20,20);     //Mouse Move rel.
       sleep(1000);
       DD_mov(100,100);    //Mouse Move abs.
     end;
     2:
     begin                    //Mouse LeftClick.
          DD_btn(1);          //1==L.down, 2==L.up, 4==R.down, 8==R.up, 16==M.down, 32==M.up
          sleep(50);          //may, delay 50ms,
          DD_btn(2);
     end;

     3:                       //Mouse Wheel. 1==down, 2==up
     begin
          DD_whl(1);
          DD_whl(2);
     end;

     4:
     begin
          ddcode := 300;              // TAB is 300 in ddcode.
          ddcode := DD_todc(VK_TAB);  //or by VK code
          DD_key(ddcode,1);           //1==down£¬2==up
          sleep(60);                  //may, delay 1-120ms,
          DD_key(ddcode,2);
     end;

     //5:  DD_str('KeyBorad !@#*| ');

   end;
end;

procedure TForm1.Button1Click(Sender: TObject);
begin
  Timer1.Enabled := true;
  Button2.Enabled:=true;
  Button1.Enabled:=false;
end;

procedure TForm1.GetDDFunAddress();
var
  st:integer;
begin
  @DD_mov:=nil;  //Mouse move abs.
  @DD_movR:=nil; //Mouse move rel.
  @DD_btn:=nil;  //Mouse button
  @DD_whl:=nil;  //Mouse wheel
  @DD_key:=nil;  //Keyboard
  @DD_str:=nil;  //Input visible char
  @DD_todc:=nil; //VK to ddcode


 if DD <> 0 then
 begin
    @DD_mov:=GetprocAddress(dd,'DD_mov');
    @DD_btn:=GetprocAddress(dd,'DD_btn');
    @DD_whl:=GetprocAddress(dd,'DD_whl');
    @DD_key:=GetprocAddress(dd,'DD_key');
    @DD_str:=GetprocAddress(dd,'DD_str');
    @DD_movR:=GetprocAddress(dd,'DD_movR');
    @DD_todc:=GetprocAddress(dd,'DD_todc');
    if ( Assigned(@DD_mov) and Assigned(@DD_btn) and
         Assigned(@DD_whl) and Assigned(@DD_key) and
         Assigned(@DD_str) and Assigned(@DD_movR)and  Assigned(@DD_todc))then
    begin
      st := DD_btn(0);//DD Initialize
      if st=1 then
      begin
        MessageBox(0,'OK','info',MB_ICONASTERISK);
        button1.Enabled := true;
      end
      else
      begin
        MessageBox(0,'Error','info',MB_ICONASTERISK);
      end;
    end;

 end;

end;

end.

